connectors/utils.py

```diff
@@ -414,6 +406,29 @@ async def put(self, coroutine, result_callback=None):
         )
         return task
 
+    async def put(self, coroutine, result_callback=None):
+        """Adds a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function will block and wait for a free slot.
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+        await self._sem.acquire()
+        return self._add_task(coroutine, result_callback=result_callback)
+
+    def try_put(self, coroutine, result_callback=None):
+        """Tries to add a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function return a None task immediately
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+
+        if self._sem.try_acquire():
+            return self._add_task(coroutine, result_callback=result_callback)
```

Comment:
I believe you want to indicate to the caller, that put failed?

E.g. raise an error or return True/False?

Reply:
Ah, I see it returns `None` in case nothing is added.

Reply:
Good point. Will do

Reply:
> Ah, I see it returns `None` in case nothing is added.

I made it explicitly return `None` in commit https://github.com/elastic/connectors/pull/1479/commits/bf54ba1364a05f74d5755778bb0da503c4c34551

I also think that `put` and `try_put` should not return the task, which should be managed internally by `ConcurrentTasks`. For all usages of `put` and `try_put`, the returned task is not used, but it's used in tests. I think it's safe to make `put` and `try_put` to return `True/False`, but I'd like to address it in a separate PR